app/services/duplicate_canonical_profile_resolver.rb

```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+# Class that deduplicates canonical profiles with identical benchmark/ref IDs
+class DuplicateCanonicalProfileResolver
+  class << self
+    def run!
+      duplicates.each do |duplicate|
+        profiles = Profile.canonical.includes(:child_profiles).where(
+          duplicate.slice(:ref_id, :benchmark_id)
+        ).order(created_at: :asc).offset(1)
```

Comment:
I'd prefer an explicit `where.not(id: duplicate)` instead of relying on the offset(1). I guess the idea here is that at least one of these profiles will survive deletion, even if it's not `duplicate`, and that's ok. Still, it seems less deterministic than explicitly excluding `duplicate`.

Reply:
The question is, which one should I preserve? My idea was that the oldest one :thinking:

Reply:
The oldest one works, but **it doesn't really matter which one we keep** if none have children. Of course, if one does have children, we'd prefer to keep that one.

Reply:
Then I don't understand your `where.not(id: duplicate)`, that would make us do +1 query as we need to fetch the ID of that "duplicate" first :confused:

Reply:
The `duplicate` record is already fetched and in memory in your `duplicates.each`. Accessing the ID or calling `duplicates.slice(...)` accesses the query result already stored in memory and does not introduce any additional queries.

Reply:
Oh, but this way we're just pushing the non-deterministic thing elsewhere, because the ID that we get there will be also non-deterministic, right?

Reply:
Also I can't ask for the ID in the select because that would mess up the group_by :confused:

Reply:
Ok yeah, both good points :laughing: